PPS Full simulation integration

mundim opened the pull request:
#### PR description:

Integration of PPS full simulation into cmssw. This is the follow up from the closed PR #28190.


#### PR validation:
scram b runtests, scram b code-checks and scram b code-format have been run again, with no issue

As for runTheMatrix, some observations need to be done:

runTheMatrix --job-reports --ibeos -j 15 -l limited -i all produced the outout

tests with errors: 
4.22, 5.1, 101.0, 135.4, 8.0, 136.7611, 136.8311, 136.88811, 140.53, 1001.0 and 1325.7

tests that failed in the first attempt, but ran ok with no change (some timeout?)
4.53, 136.731, 136.793, 136.874, 140.56 and 1000.0

tests that failed because GT run1_mc (pps transport not included) when PPS it not active:
9.0 and 25.0

The following tests due to missing PPS condition **CTPPSPixelGainCalibrationRcd** in the GT (it has been already asked to include them in the GT) (https://github.com/cms-sw/cmssw/pull/28190#issuecomment-627525205)

11634.0 and 12434.0

Tests that failed because PPS geometry from DB was not loaded (or step1.root taken from DAS, this is a file produced without PPS geometry). It has been already asked to include it in the GT  (https://github.com/cms-sw/cmssw/pull/28190#issuecomment-624283309):

7.3, 158.0, 1306.0. 1330.0, 10024.0, 10042.0, 10224.0, 10824.0, 20034.0, 20434.0, 21234.0, 21234.0, 25202.0

mundim committed:
Remaining config files needed to implement PPS simulation into cmssw (and adapting PPSPixelDigiProducer to FileInPath standard;removing old test config)

mundim committed:
Implementing PR review comments, removing unneeded Sequence usage instead of Task in the config

mundim committed:
Rebasing to the last IB; adapting test config to the new pps ERAS; removing mixNoPU_PPS_cff

mundim committed:
Modifing Generator_cff to trigger the proton propagation at generator step

mundim committed:
Removing a print statement from a test config

mundim committed:
Re-coding test config to minimize changes due to the missing DB sim geometry; Propagating changes due to renaming campaign (CTPPS->PPS)

mundim committed:
Update SimPPS/PPSPixelDigiProducer/src/RPixChargeShare.cc

Co-Authored-By: Matti Kortelainen <matti.kortelainen@cern.ch>

mundim committed:
Update Configuration/StandardSequences/python/Generator_cff.py

Co-Authored-By: Matti Kortelainen <matti.kortelainen@cern.ch>

mundim committed:
Rebasing to new IB, implementing the new proton transport

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29916/15516

- This PR adds an extra 84KB to repository

- Found files with invalid states:
  - SimGeneral/MixingModule/python/mixNoPU_PPS_cfi.py:
    - Added: e0580a0ea4786a944b8caa6e848888156741122d
    - Modified: 296b09d2c5d998077eeb7cbd6c80af97400f145a
    - Deleted: 152e64026392e8b1a4db41ec82eba162a667b176

cmsbuild commented:
A new Pull Request was created by  @mundim  for master.

It involves the following packages:

Configuration/StandardSequences
EventFilter/CTPPSRawToDigi
EventFilter/RawDataCollector
IOMC/RandomEngine
SimGeneral/MixingModule
SimPPS/Configuration
SimPPS/PPSPixelDigiProducer
SimTransport/PPSProtonTransport


@perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @emeschi, @mdhildreth, @cmsbuild, @silviodonato, @franzoni, @slava77, @mommsen, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@fabiocos, @makortel, @felicepantaleo, @forthommel, @jan-kaspar, @GiacomoSguazzoni, @rovere, @VinInn, @Martin-Grunewald, @wddgit, @lecriste, @dgulhan, @slomeo this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mundim commented:
Two more comments (@fabferro, @malbouis @wpcarvalho)

- concerning the missing PPS geometry: If I force a load of a sqlite file from a local version of Configuration/Geometry/python/GeometrySimDB_cff.py, most of them work
- as mention in the PR #28190, even loading PPS geometry, test 7.3 will fail because it uses a digi configuration for cosmic in which PPS digitization is not included and I wonder why should it be.

makortel commented on SimPPS/Configuration/python/PPSTransport_cff.py:
```diff
@@ -0,0 +1,20 @@
+g4SimHits.SteppingAction.MaxTrackTime = cms.double(2000.0)
+g4SimHits.StackingAction.MaxTrackTime = cms.double(2000.0)
+
+from IOMC.RandomEngine.IOMC_cff import *
+RandomNumberGeneratorService.LHCTransport.engineName   = cms.untracked.string('TRandom3')
```
Please remove, the `RandomNumberGeneratorService` should be fully defined in `IOMC.RandomEngin.IOMC_cff`.
```suggestion
```

makortel commented on SimPPS/Configuration/python/PPSTransport_cff.py:
```diff
@@ -0,0 +1,20 @@
+g4SimHits.Generator.MaxEtaCut = cms.double( 13.0)
+g4SimHits.Generator.HepMCProductLabel   = 'LHCTransport'
+g4SimHits.SteppingAction.MaxTrackTime = cms.double(2000.0)
+g4SimHits.StackingAction.MaxTrackTime = cms.double(2000.0)
```
Configuration fragments modifying objects defined in other files is in generally discouraged (even though we still have many files that do it). Could these be set in `g4SimHits_cfi` e.g. with the `ctpps_2016` modifier (that appears to be a general switch to turn on PPS stuff)?

mundim replied:
@makortel What do you propose? These parameters are defined in SimG4Core/Applications/python/g4SimHits_cfi.py, but I'm afraid to cause undesired impact somewhere else...

mundim replied:
Yes, we need this for all years since 2016, otherwise, PPS tracks will not be treated properly. Should I include g4SimHits_cfi in this PR then?

makortel replied:
> Should I include g4SimHits_cfi in this PR then?

Yes please.

mundim replied:
I noticed there is a modified "run2_common", since PPS is intended to be included in all run2, isn't a better idea to use this modified instead of the ctpps_[year]?

makortel replied:
The convention has been to use the "most specialized" modifier for the modifications. E.g. in this case that allows to easily construct a workflow like "Run 2 without PPS". (also strictly speaking Run 2 started in 2015, while the first `ctpps` Modifier is for 2016.)

mundim replied:
Ok.

mundim replied:
Done. It was used only ctpps_2016 (and tested against 2017 and 2018 as well) as this modified is a general one for PPS.

mundim committed:
Moving redefinition of parameters from other packages to the respective place

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29916/15575

- This PR adds an extra 104KB to repository

- Found files with invalid states:
  - SimGeneral/MixingModule/python/mixNoPU_PPS_cfi.py:
    - Added: e0580a0ea4786a944b8caa6e848888156741122d
    - Modified: 296b09d2c5d998077eeb7cbd6c80af97400f145a
    - Deleted: 152e64026392e8b1a4db41ec82eba162a667b176

cmsbuild commented:
Pull request #29916 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @emeschi, @mdhildreth, @cmsbuild, @silviodonato, @franzoni, @slava77, @mommsen, @fabiocos, @davidlange6 can you please check and sign again.

makortel commented on SimG4Core/Application/python/g4SimHits_cfi.py:
```diff
@@ -575,6 +575,15 @@
 
+
+from Configuration.Eras.Modifier_ctpps_2016_cff import ctpps_2016
+ctpps_2016.toModify( g4SimHits.Generator, HepMCProductLabel = 'LHCTransport')
+ctpps_2016.toModify( g4SimHits.Generator, MinEtaCut = cms.double(-13.0))
+ctpps_2016.toModify( g4SimHits.Generator, MaxEtaCut = cms.double( 13.0))
+ctpps_2016.toModify( g4SimHits.SteppingAction, MaxTrackTime = cms.double(2000.0))
+ctpps_2016.toModify( g4SimHits.StackingAction, MaxTrackTime = cms.double(2000.0))
```
This could also be done by
```suggestion
ctpps_2016.toModify(g4SimHits,
    Generator = dict(
        HepMCProductLabel = 'LHCTransport',
        MinEtaCut = -13.0,
        MaxEtaCut = 13.0
    ),
    SteppingAction = dict(
        MaxTrackTime = 2000.0,
        MaxTrackTime = 2000.0
    )
)
```

mundim replied:
The build error is due to an error in the suggested change in https://github.com/cms-sw/cmssw/pull/29916#discussion_r428796936.

I accept it without noticing the problem. As in the original code, the MaxTrackTime need to be changed in SteppingAction AND StackingAction. Can I change it using the web interface?

makortel replied:
Sorry, apparently I didn't look carefully enough.

mundim replied:
No problem, the words are pretty similar, I have done the same mistake the first time... I changed it by means of the web interface and submitted. Is this way ok with you? It is much straightforward.

makortel replied:
Yes, it looks fine now, thanks.

mundim committed:
Update SimG4Core/Application/python/g4SimHits_cfi.py

Co-authored-by: Matti Kortelainen <matti.kortelainen@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29916/15578

- This PR adds an extra 104KB to repository

- Found files with invalid states:
  - SimGeneral/MixingModule/python/mixNoPU_PPS_cfi.py:
    - Added: e0580a0ea4786a944b8caa6e848888156741122d
    - Modified: 296b09d2c5d998077eeb7cbd6c80af97400f145a
    - Deleted: 152e64026392e8b1a4db41ec82eba162a667b176

cmsbuild commented:
Pull request #29916 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @emeschi, @mdhildreth, @cmsbuild, @silviodonato, @franzoni, @slava77, @mommsen, @fabiocos, @davidlange6 can you please check and sign again.

perrotta commented:
@mundim please let us know when this PR can be considered for tests and review

mundim commented:
> @mundim please let us know when this PR can be considered for tests and review
From my side, it is ok to be tested, but remember that while PPS geometry is not in the GT some tests will fail. It is already in the DB, but not yet in the GT (as far as I know) and I have no control over it, we have already asked to do so. There is another problem with tags with gain calibration and mapping that were submitted by @malbouis in November but they are not in the GT either.

perrotta commented:
> >  @mundim please let us know when this PR can be considered for tests and review

> From my side, it is ok to be tested, but remember that while PPS geometry is not in the GT some tests will fail. It is already in the DB, but not yet in the GT (as far as I know) and I have no control over it, we have already asked to do so. There is another problem with tags with gain calibration and mapping that were submitted by @malbouis in November but they are not in the GT either.

Tests can be run when an appropriate GT will be available, so that they can meaninfully reach a conclusion.

If this PR is supposed to enter CMSSW at some point, someone should push for updating the GT: 
- PPS:  @jan-kaspar @fabferro 
- AlCa: @pohsun, @christopheralanwest, @tocheng, @tlampen

christopheralanwest commented:
please test

I wanted to test a GT that uses this PR locally (https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/111X_mcRun3_2021_realistic_v4/111X_mcRun3_2021_realistic_Candidate_2020_06_03_17_55_10) with the changes requested at https://github.com/cms-sw/cmssw/pull/28190#issuecomment-627525205, but I couldn't compile after merging this PR. I'm running the integration tests just to document the error.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: be785497693df0a629f3d55da599adbc5665ed9b

CMSSW: CMSSW_11_2_X_2020-06-02-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f674e2/6780/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  Python3**

* **Build**:

I found compilation warning when building:  See details on the summary page.

* **Python3**:

I found errors: 
 
 *** Error compiling 'src/SimG4Core/Application/python/g4SimHits_cfi.py'...

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

mundim commented:
Please, see the comment at https://github.com/cms-sw/cmssw/pull/29916#discussion_r434769583

mundim committed:
Update g4SimHits_cfi.py

Moving the second MaxTrackTime redefinition from SteppintAction to StackingAction.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29916/15845

- This PR adds an extra 120KB to repository

- Found files with invalid states:
  - SimGeneral/MixingModule/python/mixNoPU_PPS_cfi.py:
    - Added: e0580a0ea4786a944b8caa6e848888156741122d
    - Modified: 296b09d2c5d998077eeb7cbd6c80af97400f145a
    - Deleted: 152e64026392e8b1a4db41ec82eba162a667b176

cmsbuild commented:
Pull request #29916 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @emeschi, @mdhildreth, @cmsbuild, @silviodonato, @franzoni, @slava77, @mommsen, @fabiocos, @davidlange6 can you please check and sign again.

christopheralanwest commented:
With the new conditions that I believe that you want for the 2021 scenario:

https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/111X_mcRun3_2021_realistic_v4/111X_mcRun3_2021_realistic_Candidate_2020_06_03_17_55_10

I am still unable to successfully run a local test with a candidate GT:

```
runTheMatrix.py -l 11634.0 --ibeos --command="--conditions=111X_mcRun3_2021_realistic_Candidate_2020_06_03_17_55_10"
```

which fails with the following exception:

```
----- Begin Fatal Exception 04-Jun-2020 00:58:57 CEST-----------------------
An exception of category 'Unknown' occurred while
   [0] Processing global begin Run run: 1
   [1] Calling method for module CTPPSDiamondDQMSource/'ctppsDiamondDQMSource'
   [2] Using EventSetup component CTPPSGeometryESModule/'ctppsGeometryESModule' to make data CTPPSGeometry/'' in record VeryForwardRealGeometryRecord
   [3] Using EventSetup component CTPPSGeometryESModule/'ctppsGeometryESModule' to make data DetGeomDesc/'' in record VeryForwardRealGeometryRecord
   [4] Using EventSetup component CTPPSGeometryESModule/'ctppsGeometryESModule' to make data DetGeomDesc/'' in record IdealGeometryRecord
Exception Message:
An exception of unknown type was thrown.
----- End Fatal Exception -------------------------------------------------
```

With the tag "XMLFILE_Geometry_111YV1_2021CTPPS_mc" removed, the tests are able to complete successfully:
```
runTheMatrix.py -l 11634.0 --ibeos --command="--conditions=111X_mcRun3_2021_realistic_Candidate_2020_06_03_19_41_33"
```
Could you investigate the issue with the geometry tag?

Could you also provide a list or table that describes exactly which conditions need to be updated for each of the following scenarios:

- Run 1 MC
- 2016
- 2017
- 2018
- Run 3
- Phase 2

I suggest also that you or another PPS expert give a talk at an upcoming AlCaDB meeting (15:00 CERN time on Mondays).

Finally, I would like to suggest that you perform test of your workflows with conditions overrides to include new conditions prior to requesting that the corresponding conditions be included in a global tag. Instructions for GT customization can be found here:

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideFrontierConditions#Customization_of_Global_Tags

mundim commented:
The list is basically ready, it should be sent by @malbouis. But, PPS should not be included in any RunI scenario. Is this test (11634.0) the only one that fails or there are other ones?
Thanks.

malbouis commented:
I made a table with the tags related to simulation. It is here: https://docs.google.com/spreadsheets/d/1ewhK4F2fJoPYHUmGpqBHf0gE4ybPKddk1pHHxFAoGjc/edit?usp=sharing

mundim commented:
> `runTheMatrix.py -l 11634.0 --ibeos --command="--conditions=111X_mcRun3_2021_realistic_Candidate_2020_06_03_17_55_10"`

I have run this test (11634.0) and the problem is not with the simulation geometry. Just for the record, PPS uses different geometry for simulation and reconstruction (as other detectors, I suppose). If I use the GT candidate as above, it runs until step2, with the PPS hit collection properly created. Then, using the GT 111X_mcRun3_2021_realistic_v4, it runs step3 with no problem. There is a problem at step4, the same one that occurred when using the option reported above as "successful", but it happens with cmsDriver and doesn't seem related to the code itself.

I forgot to mention, as reported when issuing the PR, that the record (in SQLite file format) was indeed tested and worked fine.

So, my suspicion is that the PPS reco geometry was replaced by the one intended only for simulation. Could that be the case?

Thanks.

cvuosalo commented:
@malbouis The proposed tag XMLFILE_Geometry_111YV2_Extended2021CTPPS_mc looks wrong to me. Should it not be XMLFILE_CTPPS_Geometry_2021_111YV2, with label "CTPPS"? The tag "XMLFILE_Geometry..." is the entire geometry, and any modification of it needs to be coordinated with me.

mundim commented:
I tried to run the test using 111X_mcRun3_2021_realistic_Candidate_2020_06_03_19_20_01, which contains the tag XMLFILE_CTPPS_Geometry_2021_111YV2_mc, but then the PPS geometry is not loaded (the simhit collect doesn't get produced). There is no error in the output, apparently, it is just not loaded.

christopheralanwest commented:
@mundim @malbouis Carl Vuosalo will be giving a talk at Monday's AlCaDB meeting (https://indico.cern.ch/event/927125/). A PPS representative is welcome to give a talk as well but in any case I suggest attending so that we can have a discussion regarding PPS geometry tags including all of the relevant parties.

mundim commented:
> @mundim @malbouis Carl Vuosalo will be giving a talk at Monday's AlCaDB meeting (https://indico.cern.ch/event/927125/). A PPS representative is welcome to give a talk as well but in any case I suggest attending so that we can have a discussion regarding PPS geometry tags including all of the relevant parties.

I'll try to make it at least to have the discussion and clarify some (a lot perhaps) grey corners, but I might have another meeting overlapping it a bit. Thank you very much.

mundim committed:
Resolving a bug in PPSProtonTransport; moving the propagator do use db saved optical functions

mundim committed:
bringing a file compliant with code-format option

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29916/15971

- This PR adds an extra 128KB to repository

- Found files with invalid states:
  - SimGeneral/MixingModule/python/mixNoPU_PPS_cfi.py:
    - Added: e0580a0ea4786a944b8caa6e848888156741122d
    - Modified: 296b09d2c5d998077eeb7cbd6c80af97400f145a
    - Deleted: 152e64026392e8b1a4db41ec82eba162a667b176

mundim committed:
Including PPS geometry into the extended one; removing the alternative es_source from PPSTransportESSources; Adding ctpps_2021 into Run3 Era to force using the proton propagation for 2021

mundim committed:
including a remainder

cmsbuild commented:
The code-checks are being triggered in jenkins.

mundim commented:
This PR was updated in view of the discussion with @ianna, @cvuosalo and @civanch concerning the PSP geometry. In order to have the tests to complete, it is needed the new extended geometry (from cmsExtendedGeometry2021XML) to be prepared.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29916/16096

- This PR adds an extra 152KB to repository

- Found files with invalid states:
  - SimGeneral/MixingModule/python/mixNoPU_PPS_cfi.py:
    - Added: e0580a0ea4786a944b8caa6e848888156741122d
    - Modified: 296b09d2c5d998077eeb7cbd6c80af97400f145a
    - Deleted: 152e64026392e8b1a4db41ec82eba162a667b176


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Geometry/CMSCommonData/python/cmsExtendedGeometry2021XML_cfi.py  modified in PR(s): #30214

cmsbuild commented:
Pull request #29916 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @emeschi, @ianna, @mdhildreth, @cmsbuild, @silviodonato, @franzoni, @kpedro88, @slava77, @mommsen, @fabiocos, @davidlange6 can you please check and sign again.

mundim commented:
Please, put this PR on hold until PR #30214 gets merged.

christopheralanwest commented:
hold

cmsbuild commented:
Pull request has been put on hold by  @christopheralanwest
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

smuzaffar commented:
@mundim , can you please rebase this PR? Currently it has conflicts

mundim commented:
Please, don't unhold it yet. I'm just waiting for other  (already approved) PRs to appear in an IB to rebase it. In fact, I was going to ask what is the easiest way to proceed: rebase it or create a new PR?

cvuosalo commented:
@mundim Rebasing or creating a new PR are equivalent. After rebasing, you should squash the commits to compress the version history, so maybe that approach is more work. Use whichever method you find easier or more reliable.

mundim commented:
I was proceeding with this PR. In fact, I guessed it would be easier to close this one and issue another one, with clean the history. However, before doing so, I have to discuss an issue:
the runTheMatrix tests failed in almost all those ones that pick up previously simulated file (step1.root) in which PPS was not present. In test number 7.3, 158.0, and 1306.0 it does the simulation, but using the geometry DB:Extended for 2018, in which PPS is also not present. The test 21234.0 uses Extended2026D44 geometry, in which PPS is also not included.

We have agreed (@cvuosalo @civanch) that for the time being, PPS Run2 simulation should be done privately. So, is it possible to have a special testing setup that skips those tests to avoid halting the PR?
Thanks

civanch commented:
@mundim , I would suggest close this and make a new one. You should not try to change work of WFs, in which there is no PPS. As I said before PPS should be as HGCal or MTD - these Phase-2 detectors are present only in few WFs and do not break Run-2 or Run-3 WFs.

cvuosalo commented:
@mundim I agree with @civanch. A PPS PR should not interfere with existing workflows. A PR must be backwardly compatible with existing workflows and Global Tags.

mundim commented:
> @mundim , I would suggest close this and make a new one. You should not try to change work of WFs, in which there is no PPS. As I said before PPS should be as HGCal or MTD - these Phase-2 detectors are present only in few WFs and do not break Run-2 or Run-3 WFs.

Yes, this is my idea. I ran the tests in the new branch and was making you aware of the tests that failed before submitting the new PR.

mundim commented:
> @mundim I agree with @civanch. A PPS PR should not interfere with existing workflows. A PR must be backwardly compatible with existing workflows and Global Tags.

@fabferro

So, the way I see to integrate PPS simulation is either:
1 - backport the PPS geometry to Run2, including it into the extended geometry as it was done for Run3
or
2 - remove the PPS simulation from Run2, keeping the code in the CTPPS git repository (in a dedicated branch) to allow private production.

thanks

mundim commented:
I'll proceed with the option 2 above, as already decided, removing RunII PPS eras from gen and digi steps.

christopheralanwest commented:
> > @mundim I agree with @civanch. A PPS PR should not interfere with existing workflows. A PR must be backwardly compatible with existing workflows and Global Tags.

This statement about global tags is not correct. If it were, it would be impossible to introduce new records. For a CMSSW release CMSSW_X_Y_Z, the major version X and the minor version Y are encoded in the global tag name (for example, 112X for CMSSW_11_2_Z). Generally, AlCa tries to make, for example, all 112X GTs for the same scenario compatible. However, if necessary, it is possible to set the starting release also including the patch-level Z. See, for example, the "Starting Release" for [112X_mcRun3_2021_realistic_v2](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/112X_mcRun3_2021_realistic_v2) is "CMSSW_11_2_0" and so cannot be used prior to CMSSW_11_2_0 (and its prereleases). What is forbidden are any incompatibilities in patches, for example, CMSSW_X_Y_Z, patchN.

To summarize, changes like those needed for this PR would be allowed in pre-releases.

> 
> @fabferro
> 
> So, the way I see to integrate PPS simulation is either:
> 1 - backport the PPS geometry to Run2, including it into the extended geometry as it was done for Run3
> or
> 2 - remove the PPS simulation from Run2, keeping the code in the CTPPS git repository (in a dedicated branch) to allow private production.
> 
> thanks

From the AlCa perspective, I see nothing wrong with option 1) as long as the changes do not have to be backported to an earlier CMSSW release.

mundim commented:
> > > @mundim I agree with @civanch. A PPS PR should not interfere with existing workflows. A PR must be backwardly compatible with existing workflows and Global Tags.
> 
> This statement about global tags is not correct. If it were, it would be impossible to introduce new records. For a CMSSW release CMSSW_X_Y_Z, the major version X and the minor version Y are encoded in the global tag name (for example, 112X for CMSSW_11_2_Z). Generally, AlCa tries to make, for example, all 112X GTs for the same scenario compatible. However, if necessary, it is possible to set the starting release also including the patch-level Z. See, for example, the "Starting Release" for [112X_mcRun3_2021_realistic_v2](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/112X_mcRun3_2021_realistic_v2) is "CMSSW_11_2_0" and so cannot be used prior to CMSSW_11_2_0 (and its prereleases). What is forbidden are any incompatibilities in patches, for example, CMSSW_X_Y_Z, patchN.
> 
> To summarize, changes like those needed for this PR would be allowed in pre-releases.
> 
> > @fabferro
> > So, the way I see to integrate PPS simulation is either:
> > 1 - backport the PPS geometry to Run2, including it into the extended geometry as it was done for Run3
> > or
> > 2 - remove the PPS simulation from Run2, keeping the code in the CTPPS git repository (in a dedicated branch) to allow private production.
> > thanks
> 
> From the AlCa perspective, I see nothing wrong with option 1) as long as the changes do not have to be backported to an earlier CMSSW release.

So. The question is: how to do it? There was an attempt with a GT candidate with the PPS simulation geometry, but somehow, the test broke at reco level. Then, it was decided to include PPS as part of the cms extended geometry, which is done for Run3. Now, the tests are falling for Run2  in two situations:

1 - when the simulation is done in the wf but, but PPS SIM geometry is not yet in the GT, so the corresponding branch is missing in the source root file (step1.root) and digi step breaks;

2 - when the simulation is taken from a previously simulated sample which was produced without PPS, so again, the pps branch is missing and the digi step breaks.

Thanks.

cvuosalo commented:
@mundim My suggestion is to separate Run 2 from Run 3. Please take out the Run 2 changes that are causing problems in this PR, so it can go forward and eventually be merged. Then in a later PR you can try to figure out the Run 2 issues.

christopheralanwest commented:
@cvuosalo Is a DB payload for the 2021 extended geometry including PPS available yet? And are the other geometry tags described at https://github.com/cms-sw/cmssw/pull/28190#issuecomment-624283309 suitable? Inclusion of the 2018 extended geometry was sufficient to run with the previous PR; see https://github.com/cms-sw/cmssw/pull/28190#issuecomment-627468413.

@mundim I don't understand the issue with Run-2 simulation. Why are your local tests being run with the flag "-i all" to re-use the GEN-SIM? If you try to run PPS reconstruction on GEN-SIM that does not include PPS simulation, isn't failure expected (perhaps even desired)? Also, please re-base this PR.

cvuosalo commented:
@christopheralanwest The 2021 geometry DB payloads are waiting for the resolution of this PR.

Of these tags:

XMLFILE_Geometry_111YV1_2021CTPPS_mc
XMLFILE_Geometry_111YV1_Extended2016CTPPS_mc
XMLFILE_Geometry_111YV1_Extended2017CTPPS_mc
XMLFILE_Geometry_111YV1_Extended2018CTPPS_mc
XMLFILE_Geometry_111YV2_Extended2021CTPPS_mc

Only XMLFILE_Geometry_111YV1_2021CTPPS_mc seems safe because it has label "CTPPS" and thus would only be used by PPS code. XMLFILE_Geometry_111YV1_Extended2018CTPPS_mc is particularly worrisome because it has label "Extended" and thus would replace the validated 2018 Geometry File Record, with unknown consequences for 2018 simulation and reconstruction.

mundim commented:
> @mundim I don't understand the issue with Run-2 simulation. Why are your local tests being run with the flag "-i all" to re-use the GEN-SIM? If you try to run PPS reconstruction on GEN-SIM that does not include PPS simulation, isn't failure expected (perhaps even desired)? Also, please re-base this PR.

I'm following the instructions at http://cms-sw.github.io/PRWorkflow.html. I looked for a page with instructions about runTheMatrix options but I could not find it.

Yes, this PR will be re-based (or a new one created). I was sorting out the issue in this one to avoid having the same problems in the new one.

Please, confirm if I should split the PR in two, one for Run2 and another one for Run3 (I suppose it is a good idea). I'll comment further about the tags  above.

Thanks.

mundim commented:
> @christopheralanwest The 2021 geometry DB payloads are waiting for the resolution of this PR.
> 
> Of these tags:
> 
> XMLFILE_Geometry_111YV1_2021CTPPS_mc
> XMLFILE_Geometry_111YV1_Extended2016CTPPS_mc
> XMLFILE_Geometry_111YV1_Extended2017CTPPS_mc
> XMLFILE_Geometry_111YV1_Extended2018CTPPS_mc
> XMLFILE_Geometry_111YV2_Extended2021CTPPS_mc
> 
> Only XMLFILE_Geometry_111YV1_2021CTPPS_mc seems safe because it has label "CTPPS" and thus would only be used by PPS code. XMLFILE_Geometry_111YV1_Extended2018CTPPS_mc is particularly worrisome because it has label "Extended" and thus would replace the validated 2018 Geometry File Record, with unknown consequences for 2018 simulation and reconstruction.

These tags were built using XML edited by hand. XMLFILE_Geometry_111YV1_2021CTPPS_mc uses a single pps.xml, which showed some problems.

The others with CTPPS in the tag name, including 2021, as said above, use XML edited by hand. There is a new extended geometry for 2021 that uses a dictionary builder. I suppose, at least for 2021, the best thing is to upload it again using the approved (already merged) XML file.
As for 2016-2018 I don't know what to do, they have the full extended geometry including PPS and tested against PPS full simulation, but they use PPS XML files in which there are syntax errors which were allowed in the old system, but not in the new one.

civanch commented:
@mundim , @christopheralanwest , We need 2021 DB geometry, absence of DB made from recent XML slow down other works significantly, we basically stack. Before this is done there is no sense discussing Run-2. If for that an extra PR is needed, let us make it with the limited focus to Run-3.

mundim commented:
> @mundim , @christopheralanwest , We need 2021 DB geometry, absence of DB made from recent XML slow down other works significantly, we basically stack. Before this is done there is no sense discussing Run-2. If for that an extra PR is needed, let us make it with the limited focus to Run-3.

Ok, I'll start removing all Run2 stuffs right now. runTheMatrix uses to take a lot of time to process, is there a minimum option set to speed it up?

civanch commented:
@mundim , make a new PR for Run-3 - it will be much more clean and less number of files.

cvuosalo commented:
@mundim For runTheMatrix, you can first run only one or a couple of the workflows that had the most errors previously:
```
runTheMatrix --job-reports --ibeos -j 15 -l 4.22,7.3,11634.0 -i all
```
as an example.
Make sure you have a voms-proxy before calling runTheMatrix.

mundim commented:
Hi Carl. When I saw your comment, I had already submitted the test with  "-l limited" option. All of the wf except the ones below ran ok. The failed tests are (with something that called my attention):
`20034.0   -> uses geometry Extended2026D35
20434.0   -> usew geometry Extended2026D41
21234.0   -> uses geometry Extended2026D44
23234.0   -> uses geometry Extended2026D49
11634.0   -> SIM-GEN CMSSW_10_6_1-106X and missing souce for  CTPPSPixelGainCalibrationsRcd
12434.0   -> SIM-GEN CMSSW_10_6_1-106X and missing souce for  CTPPSPixelGainCalibrationsRcd
`
Are these observations relevant? I included the essource by hand (tag CTPPSPixelGainCalibrations_v1_mc), then the missing es_source disappeared.

cvuosalo commented:
@mundim You're making progress. Some of the workflows are now succeeding. You have to figure out how to get those six workflows to run without errors. I don't know what change in this PR causes wfs 11634.0 and 12434.0 to require the CTPPSPixelGainCalibrationsRcd, but that needs to be removed from this PR. For the 2026 wfs, you'll need to look through their logs to see why they had errors.

mundim commented:
> @mundim You're making progress. Some of the workflows are now succeeding. You have to figure out how to get those six workflows to run without errors. I don't know what change in this PR causes wfs 11634.0 and 12434.0 to require the CTPPSPixelGainCalibrationsRcd, but that needs to be removed from this PR. For the 2026 wfs, you'll need to look through their logs to see why they had errors.

I dug into the code and could not find anything that could cause such error. I ran a config for 2018 that should produce PPS hits if PPS was in the chain, but since I removed it, there is no PPS geometry, nor PPS hits and the job ran to completion with no problem. But when running these wfs I get the error below:

`An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 13 event: 1203 stream: 0
   [1] Running path 'HLTAnalyzerEndpath'
   [2] Prefetching for module L1TRawToDigi/'hltGtStage2Digis'
   [3] Prefetching for module RawDataCollectorByLabel/'rawDataCollector'
   [4] Prefetching for module CTPPSPixelDigiToRaw/'ctppsPixelRawData'
   [5] Calling method for module CTPPSPixelDigiProducer/'RPixDetDigitizer'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: CrossingFrame<PSimHit>
Looking for module label: mix
Looking for productInstanceName: g4SimHitsCTPPSPixelHits
`
(and also, the missing gain tag as previously mentioned).

mundim commented:
Continuing... To solve the problem with tests 20034.0, 20434.0, 21234.0 and 23234.0 I see two possibilities: or we include PPS in the Extended2026DXX or I create a new modifier for 2026 and exclude PPS.

Some of you have other suggestions? I would like to issue the new PR only after sorting this out, if you agree.
Thanks.

kpedro88 commented:
@mundim is PPS expected to be included in the Phase 2 CMS detector? if so, then it can be added to the Phase 2 geometries, but only if everything is functional. If more development will still be needed, better to exclude it for now using the `phase2_common` modifier.

christopheralanwest commented:
> @mundim You're making progress. Some of the workflows are now succeeding. You have to figure out how to get those six workflows to run without errors. I don't know what change in this PR causes wfs 11634.0 and 12434.0 to require the CTPPSPixelGainCalibrationsRcd, but that needs to be removed from this PR. For the 2026 wfs, you'll need to look through their logs to see why they had errors.

The Run-3 GTs (for example, [112X_mcRun3_2021_realistic_v2](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/112X_mcRun3_2021_realistic_v2)) do not include a `CTPPSPixelGainCalibrationsRcd` but the 2018 GTs (for example, [111X_upgrade2018_realistic_v1](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/111X_upgrade2018_realistic_v1)) do. This is a historical accident (the Run-3 queues were initially created with 2017 conditions).

Normally AlCa requires that physics validation has been performed using new conditions prior to putting them in a global tag. The problem here is that the combined code+conditions doesn't run without crashing so AlCa cannot provide an updated global tag with all of the updated conditions.

However, since the PPS simulation is not yet included in CMSSW, I think that the PPS-specific tags--all tags except the SIM geometry with the `Extended` label could be included in a GT update and their content would simply be ignored. Is that correct? If so, AlCa could provide GT updates that include these tags in a separate PR that, once merged, would eliminate some of the spurious technical issues associated with missing tags.

makortel commented:
> If more development will still be needed, better to exclude it for now using the `phase2_common` modifier.

Perhaps excluding the (relevant) PPS modifiers from `Phase2` era
https://github.com/cms-sw/cmssw/blob/2161470ec9b421f6b73327d0f0ec4d95a7f404a9/Configuration/Eras/python/Era_Phase2_cff.py#L19
would be simpler.

mundim commented:
> @mundim is PPS expected to be included in the Phase 2 CMS detector? if so, then it can be added to the Phase 2 geometries, but only if everything is functional. If more development will still be needed, better to exclude it for now using the `phase2_common` modifier.

Yes, but much more development will be needed. New stations will be installed, for which we don't have the full description yet and any attempt to propagate the geometry for phase 2 would delay even more the run3 integration. I prefer to exclude PPS from phase 2 now and work on a more permanent solutino afterwards.

kpedro88 commented:
@mundim I agree, prioritize the Run 3 integration. When you remove PPS from phase 2, include a comment to note that the removal is temporary.

mundim commented:
> > @mundim You're making progress. Some of the workflows are now succeeding. You have to figure out how to get those six workflows to run without errors. I don't know what change in this PR causes wfs 11634.0 and 12434.0 to require the CTPPSPixelGainCalibrationsRcd, but that needs to be removed from this PR. For the 2026 wfs, you'll need to look through their logs to see why they had errors.
> 
> The Run-3 GTs (for example, [112X_mcRun3_2021_realistic_v2](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/112X_mcRun3_2021_realistic_v2)) do not include a `CTPPSPixelGainCalibrationsRcd` but the 2018 GTs (for example, [111X_upgrade2018_realistic_v1](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/111X_upgrade2018_realistic_v1)) do. This is a historical accident (the Run-3 queues were initially created with 2017 conditions).
> 
> Normally AlCa requires that physics validation has been performed using new conditions prior to putting them in a global tag. The problem here is that the combined code+conditions doesn't run without crashing so AlCa cannot provide an updated global tag with all of the updated conditions.
> 
> However, since the PPS simulation is not yet included in CMSSW, I think that the PPS-specific tags--all tags except the SIM geometry with the `Extended` label could be included in a GT update and their content would simply be ignored. Is that correct? If so, AlCa could provide GT updates that include these tags in a separate PR that, once merged, would eliminate some of the spurious technical issues associated with missing tags.

This looks like a solution, but just to be clear, the correct tags are CTPPSPixelGainCalibrations_v1_mc, CTPPSPixelAnalysisMask_v1_mc and CTPPSPixelDAQMapping_v1_mc, which is already in the DB as announced by @malbouis .

mundim commented:
> > If more development will still be needed, better to exclude it for now using the `phase2_common` modifier.
> 
> Perhaps excluding the (relevant) PPS modifiers from `Phase2` era
> 
> https://github.com/cms-sw/cmssw/blob/2161470ec9b421f6b73327d0f0ec4d95a7f404a9/Configuration/Eras/python/Era_Phase2_cff.py#L19
> 
> would be simpler.

I have included PPS into Run3 modifier, not run3_common. Have I done it properly or do you have a better way? I mean, I modified Configuration.Eras.Era_Run3_cff.py with:

`
from Configuration.Eras.Modifier_ctpps_2021_cff import ctpps_2021

Run3 = cms.ModifierChain(Run2_2018.copyAndExclude([run2_GEM_2017]), run3_common, run3_GEM, run3_HB, stage2L1Trigger_2021,ctpps_2021)
`

makortel commented:
> > > If more development will still be needed, better to exclude it for now using the `phase2_common` modifier.
> > 
> > 
> > Perhaps excluding the (relevant) PPS modifiers from `Phase2` era
> > https://github.com/cms-sw/cmssw/blob/2161470ec9b421f6b73327d0f0ec4d95a7f404a9/Configuration/Eras/python/Era_Phase2_cff.py#L19
> > 
> > would be simpler.
> 
> I have included PPS into Run3 modifier, not run3_common. Have I done it properly or do you have a better way? I mean, I modified Configuration.Eras.Era_Run3_cff.py with:
> 
> `from Configuration.Eras.Modifier_ctpps_2021_cff import ctpps_2021`
> ` `
> `Run3 = cms.ModifierChain(Run2_2018.copyAndExclude([run2_GEM_2017]), run3_common, run3_GEM, run3_HB, stage2L1Trigger_2021,ctpps_2021)`

That looks correct. I meant that to remove PPS from Phase2, it could be simpler to include `ctpps_2021` modifier to the list of modifiers excluded when copying Run3 into Phase2, than to start adding `phase2_common` everywhere with exclusion purposes.

kpedro88 commented:
@makortel indeed, that is probably simpler.

mundim commented:
So, I implemented what was discussed above, removing PPS from phase2. Now ALL tests run to completion excepting 11634.0 and 12434.0 because of the CTPPSPixelGainCalibrationsRcd as above. I tried to include an es_source by hand in the step2 config file, but then there is a crash due to segmentation violation.

mundim commented:
Hello. Can I  close this PR and issue a new one, even with the problems I mentioned above with wf 11634.0 and 12434.0?

Thanks.

mundim commented:
Just one more information: I tried both tests above in a clean area, NO change at all, using the IB of yesterday, 11:00, but using the command:
`runTheMatrix.py --job-reports -j 14 --ibeos -i all -l 11634.0,12434.0  --command="--conditions=111X_upgrade2018_realistic_v1"`

As said by @christopheralanwest, this GT candidate is supposed to have the PPS gains and the same crash happened (it was just a try, I'm not sure this makes sense).

cvuosalo commented:
@christopheralanwest @mundim I think the next step is to update the GT for the PPS-specific tags mentioned here: https://github.com/cms-sw/cmssw/pull/29916#issuecomment-652606044
If I understand correctly, those tags should allow all the failing workflows to run successfully.
Then @mundim could create a new PR based upon this one for Run 3 only.

mundim commented:
> @christopheralanwest @mundim I think the next step is to update the GT for the PPS-specific tags mentioned here: [#29916 (comment)](https://github.com/cms-sw/cmssw/pull/29916#issuecomment-652606044)
> If I understand correctly, those tags should allow all the failing workflows to run successfully.
> Then @mundim could create a new PR based upon this one for Run 3 only.

I'm just waiting for a "go ahead" to create the new PR... Thanks.

perrotta commented:
-1
(As far as I understand, a clean new PR will have to be created based on this one, or alternatively a rebase of this PR is in order. While waiting for things proceeding somehow, let take this PR out from those routinely kept monitored by reco)

mundim commented:
> -1
> (As far as I understand, a clean new PR will have to be created based on this one, or alternatively a rebase of this PR is in order. While waiting for things proceeding somehow, let take this PR out from those routinely kept monitored by reco)

Yes, I prefer to create a clean/new PR. I thought it would be a good idea to keep this one open until we have everything sorted out, but if you think it is better to close it, I have nothing against it.

mundim commented:
Dear all. I'll close this PR and issue the other one. Any problem with the tag eventually not yet in the GT we discuss in the new PR, ok? Thanks.
